File: app/src/main/java/androidx/preference/FontManager.java
// Copyright (C) 2011 George Yunaev @ Ulduzsoft
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

package androidx.preference;

import android.content.Context;
import android.graphics.Typeface;
import android.os.Environment;

import androidx.annotation.NonNull;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

class FontManager {
    static List<String> getFontSearchDirectories(Context context) {
        List<String> out = new ArrayList<>(Arrays.asList("/system/fonts",
                Environment.getExternalStorageDirectory().toString() + "/weechat"));
        File appSpecificFontFolder = context.getExternalFilesDir("fonts");
        if (appSpecificFontFolder != null)
            out.add(appSpecificFontFolder.toString());
        return out;
    }

    static @NonNull LinkedList<FontInfo> enumerateFonts(Context context) {
        LinkedList<FontInfo> fonts = new LinkedList<>();

        for (String directory : getFontSearchDirectories(context)) {
            File dir = new File(directory);
            if (!dir.exists()) continue;
            File[] files = dir.listFiles();
            if (files == null) continue;

            for (File file : files) {
                String fileName = file.getName().toLowerCase();
                if (fileName.endsWith(".ttf") || fileName.endsWith(".otf")) {
                    try {
                        Typeface typeface = Typeface.createFromFile(file.getAbsolutePath());
                        fonts.add(new FontInfo(file.getName(), file.getAbsolutePath(), typeface));
                    } catch (RuntimeException r) {
                        // Invalid font
                    }
                }
            }
        }
        return fonts;
    }

    static public class FontInfo implements Comparable<FontInfo> {
        FontInfo(@NonNull String name, @NonNull String path, @NonNull Typeface typeface) {
            this.name = name;
            this.path = path;
            this.typeface = typeface;
        }

        public @NonNull String name;
        public @NonNull String path;
        public @NonNull Typeface typeface;

        @Override public int compareTo(@NonNull FontInfo another) {
            return this.name.compareTo(another.name);
        }
    }
}

File: app/src/main/java/androidx/preference/FontManager.kt
// Copyright (C) 2011 George Yunaev @ Ulduzsoft
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
package androidx.preference

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import com.ubergeek42.WeechatAndroid.R
import com.ubergeek42.WeechatAndroid.upload.Suri
import com.ubergeek42.WeechatAndroid.upload.resolver
import com.ubergeek42.WeechatAndroid.upload.suppress
import com.ubergeek42.WeechatAndroid.utils.Toaster
import com.ubergeek42.WeechatAndroid.utils.getUris
import com.ubergeek42.WeechatAndroid.utils.saveUriToFile
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import java.io.File


internal object FontManager {
    fun getFontSearchDirectories(context: Context): List<String> {
        val internalFontFolder = context.getExternalFilesDir(CUSTOM_FONTS_DIRECTORY)
        return if (internalFontFolder != null) {
            listOf("/system/fonts", internalFontFolder.toString())
        } else {
            listOf("/system/fonts")
        }
    }

    fun enumerateFonts(context: Context): List<FontInfo> {
        val fonts = mutableListOf<FontInfo>()

        getFontSearchDirectories(context).forEach { directoryName ->
            val directory = File(directoryName)
            if (directory.exists()) {
                directory.listFiles()?.forEach { file ->
                    suppress<Exception>(showToast = true) {
                        val typeface = Typeface.createFromFile(file.absolutePath)
                        fonts.add(FontInfo(file.name, file.absolutePath, typeface))
                    }
                }
            }
        }

        return fonts
    }

    fun requestFontImport(activity: Activity) {
        activity.startActivityForResult(importFontsIntent, IMPORT_FONTS_REQUEST_CODE, null)
    }

    @JvmStatic fun importFontsFromResultIntent(context: Context, intent: Intent?) {
        val imported = mutableListOf<String>()

        intent?.getUris()?.forEach { uri ->
            suppress<Exception>(showToast = true) {
                val mediaType = resolver.getType(uri)?.toMediaTypeOrNull()
                val fileName = Suri.makeFileNameWithExtension(uri, mediaType)
                val folder = context.getExternalFilesDir(CUSTOM_FONTS_DIRECTORY)
                val file = File(folder, fileName)

                try {
                    context.saveUriToFile(uri, file)
                } catch (e: Exception) {
                    file.delete()
                    throw e
                }

                // Typeface.createFromFile() does not validate the font,
                // so we are not doing any validation

                imported.add(fileName)
            }
        }

        if (imported.isNotEmpty()) {
            Toaster.SuccessToast.show(context.getString(R.string.pref__FontPreference__imported,
                                                        imported.joinToString(", ")))
        }
    }
}


internal class FontInfo (
    var name: String,
    var path: String,
    var typeface: Typeface
)


const val CUSTOM_FONTS_DIRECTORY = "fonts"
const val IMPORT_FONTS_REQUEST_CODE = 1234


private val importFontsIntent = Intent(Intent.ACTION_GET_CONTENT).apply {
    type = "font/*"
    flags = Intent.FLAG_GRANT_READ_URI_PERMISSION or
            Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION
    addCategory(Intent.CATEGORY_OPENABLE)
    putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
}